public/js/ace-1.2.5/worker-latex.js

```diff
@@ -1414,67 +1414,412 @@ var Mirror = require("../worker/mirror").Mirror;
+	    } else {
+		Tokens.push([lineNumber, code, pos, nextSpecialPos, text.slice(idx, nextSpecialPos)]);
+		var char;
+		while ((char = text[nextSpecialPos]) === ' ' || char === '\t' || char  === '\r' || char === '\n') {
+		    nextSpecialPos++;
+		    if (char === '\n') { lineNumber++; linePosition[lineNumber] = nextSpecialPos;};
 		}
+		idx = SPECIAL.lastIndex = nextSpecialPos;
+	    }
+	} else if (code === "{") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "}") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "$") {
+	    if (text[idx] === "$") {
+		idx = SPECIAL.lastIndex = idx + 1;
+		Tokens.push([lineNumber, "$$", pos]);
+	    } else {
+		Tokens.push([lineNumber, code, pos]);
+	    }
+	} else if (code === "&") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "#") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "^") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "_") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else if (code === "~") {
+	    Tokens.push([lineNumber, code, pos]);
+	} else {
+	    throw "unrecognised character " + code;
+	}
+    }
+
+    var read1arg = function (k) {
+	var open = Tokens[k+1];
+	var env = Tokens[k+2];
+	var close = Tokens[k+3];
+	var envName;
+
+	if(open && open[1] === "\\") {
+	    envName = open[4];
+	    return k + 1;
+	} else if(open && open[1] === "{" && env && env[1] === "\\" && close && close[1] === "}") {
+	    envName = env[4];
+	    return k + 3;
+	} else {
+	    return null;
+	}
+    };
+
+    var read1name = function (k) {
+	var open = Tokens[k+1];
+	var env = Tokens[k+2];
+	var close = Tokens[k+3];
+
+	if(open && open[1] === "{" && env && env[1] === "Text" && close && close[1] === "}") {
+	    var envName = text.substring(env[2], env[3]);
+	    return k + 3;
+	} else {
+	    return null;
+	}
+    };
+
+
+
+    var readOptionalParams = function(k) {
+	var params = Tokens[k+1];
+
+	if(params && params[1] === "Text") {
+	    var paramNum = text.substring(params[2], params[3]);
+	    if (paramNum.match(/^\[\d+\]$/)) {
+		return k + 1;
 	    };
 	};
-	if (error) {
-	    errors.push(error);
+	return null;
+    };
+
+    var readDefinition = function(k) {
+	k = k + 1;
+	var count = 0;
+	var nextToken = Tokens[k];
+	while (nextToken && nextToken[1] === "Text") {
+	    var start = nextToken[2], end = nextToken[3];
+	    for (i = start; i < end; i++) {
+		var char = text[i];
+		if (char === ' ' || char === '\t' || char  === '\r' || char === '\n') { continue; }
+		return null;
+	    }
+	    k++;
+	    nextToken = Tokens[k];
+	}
+	if (nextToken && nextToken[1] === "{") {
+	    count++;
+	    while (count>0) {
+		k++;
+		nextToken = Tokens[k];
+		if(!nextToken) { break; };
+		if (nextToken[1] === "}") { count--; }
+		if (nextToken[1] === "{") { count++; }
+	    }
+	    return k;
+	}
+	return null;
+    };
+
+    for (var _j = 0, _len = Tokens.length; _j < _len; _j++) {
```

Comment:
Its a little confusing that we have gone main event loop -> Helper functions, then into some sort of other processing? If this block of code was wrapped in a named function it would be clearer what it does